canonical/operator
# Add logging for custom events

ops does not log when the Framework is emitting a custom event - i.e. an event that is neither the 'main' event being fired by juju that is causing the charm to run, nor a deferred event now being reemitted.

This PR adds a debug logline: `"Emitting custom event <event instance>"`.

(This is also necessary for tools tailing the logs such as `jhack tail` to track custom events.)

## QA steps

I have added utests that verify that the logline is printed and formatted as expected, and it plays nice with the existing testing machinery.

I'm not convinced that the place/way in which I am determining whether the event is custom or not is the right one.

Perhaps we should get rid of the logger call in `ops/main.py` and consolidate the logging in Framework: add a `_log_event(event:EventBase)` method that, based on the type of the event and on the deferral status, emits the appropriate logline. Suggestions welcome.

## Documentation changes

This should not affect documentation.

## Bug reference

#842 

## Changelog

- `ops` now logs custom events.
